File: app/src/main/java/com/mxt/anitrend/base/custom/sheet/BottomSheetBase.java
package com.mxt.anitrend.base.custom.sheet;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.design.widget.BottomSheetBehavior;
import android.support.design.widget.BottomSheetDialogFragment;
import android.support.design.widget.CoordinatorLayout;
import android.support.v7.widget.AppCompatImageView;
import android.util.Log;
import android.view.View;

import com.miguelcatalan.materialsearchview.MaterialSearchView;
import com.mxt.anitrend.R;
import com.mxt.anitrend.base.custom.view.text.SingleLineTextView;
import com.mxt.anitrend.base.custom.viewmodel.ViewModelBase;
import com.mxt.anitrend.base.interfaces.event.BottomSheetChoice;
import com.mxt.anitrend.base.interfaces.event.BottomSheetListener;
import com.mxt.anitrend.base.interfaces.event.ResponseCallback;
import com.mxt.anitrend.presenter.base.BasePresenter;
import com.mxt.anitrend.util.CompatUtil;
import com.mxt.anitrend.util.KeyUtil;

import org.greenrobot.eventbus.EventBus;

import butterknife.BindView;
import butterknife.Unbinder;

/**
 * Created by max on 2017/11/02.
 * Custom bottom sheet base implementation
 */

public abstract class BottomSheetBase<T> extends BottomSheetDialogFragment implements BottomSheetListener, ResponseCallback {

    public String TAG;

    protected Unbinder unbinder;

    protected @BindView(R.id.toolbar_title) SingleLineTextView toolbarTitle;
    protected @BindView(R.id.toolbar_state) AppCompatImageView toolbarState;
    protected @BindView(R.id.toolbar_search) AppCompatImageView toolbarSearch;
    protected @BindView(R.id.search_view) MaterialSearchView searchView;

    protected ViewModelBase<T> viewModel;
    protected BottomSheetChoice bottomSheetChoice;

    protected @StringRes int mTitle, mText, mPositive, mNegative;
    protected String searchQuery;

    protected BasePresenter presenter;
    protected BottomSheetBehavior bottomSheetBehavior;
    protected BottomSheetBehavior.BottomSheetCallback bottomSheetCallback = new BottomSheetBehavior.BottomSheetCallback() {
        @Override
        public void onStateChanged(@NonNull View bottomSheet, @BottomSheetBehavior.State int newState) {
            if(isAlive()) {
                try {
                    switch (newState) {
                        case BottomSheetBehavior.STATE_HIDDEN:
                            dismiss();
                            break;
                        case BottomSheetBehavior.STATE_COLLAPSED:
                            onStateCollapsed();
                            break;
                        case BottomSheetBehavior.STATE_EXPANDED:
                            onStateExpanded();
                            break;
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }

        @Override
        public void onSlide(@NonNull View bottomSheet, float slideOffset) {

        }
    };

    /**
     * Set up your custom bottom sheet and check for arguments if any
     */
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        TAG = this.toString();
        super.onCreate(savedInstanceState);
        if(getArguments() != null) {
            mTitle = getArguments().getInt(KeyUtil.arg_title);
            mText = getArguments().getInt(KeyUtil.arg_text);
            mPositive = getArguments().getInt(KeyUtil.arg_positive_text);
            mNegative = getArguments().getInt(KeyUtil.arg_negative_text);
        }
        presenter = new BasePresenter(getContext());
    }

    /**
     * Setup your view un-binder here as well as inflating other views as needed
     * into your view stub
     */
    @Override @NonNull
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        return super.onCreateDialog(savedInstanceState);
    }

    @Override
    public void onStart() {
        super.onStart();
        toolbarTitle.setText(mTitle);
        if(bottomSheetBehavior.getState() == BottomSheetBehavior.STATE_EXPANDED)
            toolbarState.setImageDrawable(CompatUtil.INSTANCE.getTintedDrawable(getContext(), R.drawable.ic_keyboard_arrow_down_grey_600_24dp));
        else
            toolbarState.setImageDrawable(CompatUtil.INSTANCE.getTintedDrawable(getContext(), R.drawable.ic_close_grey_600_24dp));
        toolbarState.setOnClickListener(view -> {
            switch (bottomSheetBehavior.getState()) {
                case BottomSheetBehavior.STATE_EXPANDED:
                    bottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);
                    break;
                default:
                    bottomSheetBehavior.setState(BottomSheetBehavior.STATE_HIDDEN);
                    break;
            }
        });
        toolbarSearch.setImageDrawable(CompatUtil.INSTANCE.getTintedDrawable(getContext(), R.drawable.ic_search_grey_600_24dp));
        toolbarSearch.setOnClickListener(view -> searchView.showSearch(true));
        searchView.setCursorDrawable(R.drawable.material_search_cursor);
    }

    @Override
    public void onStop() {
        if(EventBus.getDefault().isRegistered(this))
            EventBus.getDefault().unregister(this);
        super.onStop();
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
    }

    public void setBottomSheetChoice(BottomSheetChoice bottomSheetChoice) {
        this.bottomSheetChoice = bottomSheetChoice;
    }

    protected boolean isAlive() {
        return isVisible() && !isDetached() || !isRemoving();
    }

    protected void createBottomSheetBehavior(View contentView) {
        CoordinatorLayout.LayoutParams layoutParams = (CoordinatorLayout.LayoutParams) ((View) contentView.getParent()).getLayoutParams();
        CoordinatorLayout.Behavior coordinatorBehavior = layoutParams.getBehavior();

        if (coordinatorBehavior instanceof BottomSheetBehavior) {
            bottomSheetBehavior = (BottomSheetBehavior) coordinatorBehavior;
            bottomSheetBehavior.setPeekHeight(CompatUtil.INSTANCE.dipToPx(KeyUtil.PEEK_HEIGHT));
            bottomSheetBehavior.setBottomSheetCallback(bottomSheetCallback);
        }
    }

    public boolean closeDialog() {
        if(bottomSheetBehavior.getState() != BottomSheetBehavior.STATE_HIDDEN) {
            bottomSheetBehavior.setState(BottomSheetBehavior.STATE_HIDDEN);
            return true;
        }
        return false;
    }

    /**
     * Remove dialog.
     */
    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if(unbinder != null)
            unbinder.unbind();
        bottomSheetCallback = null;
    }

    @Override
    public void onStateCollapsed() {
        toolbarState.setImageDrawable(CompatUtil.INSTANCE.getTintedDrawable(getContext(), R.drawable.ic_close_grey_600_24dp));
    }

    @Override
    public void onStateExpanded() {
        toolbarState.setImageDrawable(CompatUtil.INSTANCE.getTintedDrawable(getContext(), R.drawable.ic_keyboard_arrow_down_grey_600_24dp));
    }

    /**
     * Builder class for bottom sheet
     */
    public abstract static class BottomSheetBuilder {

        protected Bundle bundle;

        public BottomSheetBuilder() {
            bundle = new Bundle();
        }

        public abstract BottomSheetBase build();

        public BottomSheetBase buildWithCallback(BottomSheetChoice bottomSheetChoice) {
            BottomSheetBase bottomSheetBase = build();
            bottomSheetBase.setBottomSheetChoice(bottomSheetChoice);
            return bottomSheetBase;
        }

        public BottomSheetBuilder setTitle(@StringRes int title) {
            bundle.putInt(KeyUtil.arg_title, title);
            return this;
        }

        public BottomSheetBuilder setPositiveText(@StringRes int positiveText) {
            bundle.putInt(KeyUtil.arg_positive_text, positiveText);
            return this;
        }

        public BottomSheetBuilder setNegativeText(@StringRes int negativeText) {
            bundle.putInt(KeyUtil.arg_negative_text, negativeText);
            return this;
        }
    }

    @Override
    public void showError(String error) {
        Log.e(TAG, error);
    }

    @Override
    public void showEmpty(String message) {
        Log.d(TAG, message);
    }
}

File: app/src/main/java/com/mxt/anitrend/base/custom/sheet/BottomSheetBase.kt
package com.mxt.anitrend.base.custom.sheet

import android.app.Dialog
import android.os.Bundle
import android.view.View
import android.widget.ImageButton
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.lifecycle.Lifecycle
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.mxt.anitrend.R
import com.mxt.anitrend.base.custom.view.search.MaterialSearchView
import com.mxt.anitrend.base.interfaces.event.BottomSheetChoice
import com.mxt.anitrend.base.interfaces.event.BottomSheetListener
import com.mxt.anitrend.base.interfaces.event.ISearchDelegate
import com.mxt.anitrend.base.interfaces.event.ResponseCallback
import com.mxt.anitrend.extension.getCompatTintedDrawable
import com.mxt.anitrend.presenter.base.BasePresenter
import com.mxt.anitrend.util.KeyUtil
import org.greenrobot.eventbus.EventBus
import org.koin.android.ext.android.inject
import timber.log.Timber

/**
 * Created by max on 2017/11/02.
 * Custom bottom sheet base implementation
 */
abstract class BottomSheetBase<T> :
    BottomSheetDialogFragment(),
    BottomSheetListener,
    ResponseCallback {

    protected var toolbarTitle: TextView? = null
    protected var toolbarState: ImageButton? = null
    protected var toolbarSearch: ImageButton? = null
    protected var searchView: MaterialSearchView? = null

    protected var mSearchDelegate: ISearchDelegate? = null

    protected var bottomSheetChoice: BottomSheetChoice? = null

    @StringRes
    protected var mTitle: Int = 0

    @StringRes
    protected var mText: Int = 0

    @StringRes
    protected var mPositive: Int = 0

    @StringRes
    protected var mNegative: Int = 0
    protected var searchQuery: String? = null

    protected val presenter by inject<BasePresenter>()
    protected var bottomSheetBehavior: BottomSheetBehavior<*>? = null
    protected var bottomSheetCallback: BottomSheetBehavior.BottomSheetCallback? =
        object : BottomSheetBehavior.BottomSheetCallback() {
            override fun onStateChanged(bottomSheet: View, newState: Int) {
                if (!lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED)) {
                    return
                }
                try {
                    when (newState) {
                        BottomSheetBehavior.STATE_HIDDEN -> dismiss()
                        BottomSheetBehavior.STATE_COLLAPSED -> onStateCollapsed()
                        BottomSheetBehavior.STATE_EXPANDED -> onStateExpanded()
                    }
                } catch (e: Exception) {
                    Timber.e(e)
                }
            }

            override fun onSlide(bottomSheet: View, slideOffset: Float) = Unit
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // Toolbar presentation state (string-resource ids), not navigation identity:
        // intentionally kept on the legacy bundle channel like other presentation reads.
        arguments?.let { args ->
            mTitle = args.getInt(KeyUtil.arg_title)
            mText = args.getInt(KeyUtil.arg_text)
            mPositive = args.getInt(KeyUtil.arg_positive_text)
            mNegative = args.getInt(KeyUtil.arg_negative_text)
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog = super.onCreateDialog(savedInstanceState)

    override fun onStart() {
        super.onStart()
        toolbarTitle?.setText(mTitle)
        val ctx = context
        val closeIcon = ctx?.getCompatTintedDrawable(R.drawable.ic_close_grey_600_24dp)
        bottomSheetBehavior?.apply {
            skipCollapsed = true
            state = BottomSheetBehavior.STATE_EXPANDED
        }
        toolbarState?.setImageDrawable(closeIcon)
        toolbarState?.setOnClickListener {
            closeDialog()
        }
        toolbarSearch?.setImageDrawable(ctx?.getCompatTintedDrawable(R.drawable.ic_search_grey_600_24dp))
        toolbarSearch?.setOnClickListener {
            searchView?.showSearch()
        }
        searchView?.setOnQueryTextListener(object : MaterialSearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?): Boolean {
                mSearchDelegate?.onSearchSubmitted(query)
                return true
            }
            override fun onQueryTextChange(newText: String?): Boolean {
                mSearchDelegate?.onQueryChanged(newText)
                return false
            }
        })
        searchView?.setOnSearchViewListener(object : MaterialSearchView.SearchViewListener {
            override fun onSearchViewShown() {
                mSearchDelegate?.onSearchShown()
            }
            override fun onSearchViewClosed() {
                mSearchDelegate?.onSearchClosed()
            }
        })
    }

    override fun onStop() {
        if (EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().unregister(this)
        }
        super.onStop()
    }

    protected fun createBottomSheetBehavior(contentView: View) {
        val parentView = contentView.parent as? View ?: return
        val layoutParams = parentView.layoutParams as? CoordinatorLayout.LayoutParams ?: return
        val coordinatorBehavior = layoutParams.behavior as? BottomSheetBehavior<*> ?: return

        bottomSheetBehavior = coordinatorBehavior
        bottomSheetBehavior?.skipCollapsed = true
        bottomSheetBehavior?.isHideable = true
        bottomSheetCallback?.let { callback ->
            bottomSheetBehavior?.addBottomSheetCallback(callback)
        }
    }

    protected fun bindToolbarViews(rootView: View) {
        toolbarTitle = rootView.findViewById(R.id.toolbar_title)
        toolbarState = rootView.findViewById(R.id.toolbar_state)
        toolbarSearch = rootView.findViewById(R.id.toolbar_search)
    }

    fun closeDialog(): Boolean {
        if (bottomSheetBehavior?.state != BottomSheetBehavior.STATE_HIDDEN) {
            bottomSheetBehavior?.state = BottomSheetBehavior.STATE_HIDDEN
            return true
        }
        return false
    }

    override fun onDestroyView() {
        bottomSheetCallback?.let { callback ->
            bottomSheetBehavior?.removeBottomSheetCallback(callback)
        }
        super.onDestroyView()
        bottomSheetCallback = null
    }

    override fun onStateCollapsed() {
        toolbarState?.setImageDrawable(
            context?.getCompatTintedDrawable(R.drawable.ic_close_grey_600_24dp),
        )
    }

    override fun onStateExpanded() {
        toolbarState?.setImageDrawable(context?.getCompatTintedDrawable(R.drawable.ic_close_grey_600_24dp))
    }

    abstract class BottomSheetBuilder {
        protected val bundle: Bundle = Bundle()

        abstract fun build(): BottomSheetBase<*>

        fun buildWithCallback(bottomSheetChoice: BottomSheetChoice): BottomSheetBase<*> {
            val bottomSheetBase = build()
            bottomSheetBase.bottomSheetChoice = bottomSheetChoice
            return bottomSheetBase
        }

        fun setTitle(@StringRes title: Int): BottomSheetBuilder {
            bundle.putInt(KeyUtil.arg_title, title)
            return this
        }

        fun setPositiveText(@StringRes positiveText: Int): BottomSheetBuilder {
            bundle.putInt(KeyUtil.arg_positive_text, positiveText)
            return this
        }

        fun setNegativeText(@StringRes negativeText: Int): BottomSheetBuilder {
            bundle.putInt(KeyUtil.arg_negative_text, negativeText)
            return this
        }
    }

    override fun showError(error: String) {
        Timber.e(error)
    }

    override fun showEmpty(message: String) {
        Timber.d(message)
    }
}
